### PR TITLE
Feature/shared lock

### PR DESCRIPTION
**Pull request by kavics**
Please review the full feature. Maybe per commit review is the easier way.

### CODE REVIEW

File: src/Services/Wopi/WopiNodeOperationValidator.cs [hunk 1]
```diff
@@ -0,0 +1,63 @@
+﻿using System;
```
Comment (tusmester) [inline src/Services/Wopi/WopiNodeOperationValidator.cs]: I think this validator should be in the repo layer.

File: src/Services/Wopi/WopiService.cs [hunk 2]
```diff
@@ -0,0 +1,39 @@
+﻿using System.Collections.Generic;
+using System.Linq;
+using SenseNet.Configuration;
+using SenseNet.ContentRepository.Storage;
+
+namespace SenseNet.Services.Wopi
+{
+    internal class WopiService : ISnService
+    {
+        public static readonly string ExpectedSharedLock = "WOPI_ExpectedSharedLock";
```
Comment (tusmester) [inline src/Services/Wopi/WopiService.cs]: I think this constant should be moved to the repo layer.

File: src/Storage/SharedLock.cs [hunk 3]
```diff
@@ -43,5 +44,11 @@ public static string Unlock(int contentId, string @lock)
         {
             return Storage.DeleteSharedLock(contentId, @lock);
         }
+
+        public static void Cleanup()
+        {
+            SnTrace.Write("Cleanup shared locks.");
```
Comment (tusmester) [inline src/Storage/SharedLock.cs]: This line will write to the `Custom` trace category. We should use one of the built-in categories.

File: src/Services/Wopi/WopiHandler.cs [hunk 4]
```diff
@@ -235,25 +236,26 @@ private UserPermissions GetUserPermissions(File file, IUser user)
...
                 return new UserPermissions
                 {
-                    Write = false,
+                    Write = (allowBits & PermissionType.Save.Mask) > 0,
                     RestrictedViewOnly = 0 == (allowBits & PermissionType.Open.Mask) &&
                                          0 != (allowBits & (PermissionType.Preview.Mask +
                                                             PermissionType.PreviewWithoutWatermark.Mask +
                                                             PermissionType.PreviewWithoutRedaction.Mask)),
-                    Create = false,
+                    Create = !file.Parent?.Security.HasPermission(user, PermissionType.AddNew) ?? false,
```
Comment (tusmester) [inline src/Services/Wopi/WopiHandler.cs]: Why negate the value?

File: src/Storage/SharedLock.cs (hunk 3 above)
Comment (kavics) [reply src/Storage/SharedLock.cs]: Fixed.

File: src/Services/Wopi/WopiHandler.cs (hunk 4 above)
Comment (kavics) [reply src/Services/Wopi/WopiHandler.cs]: This is wrong. Changed to `false` constant.

**Pull request merged**